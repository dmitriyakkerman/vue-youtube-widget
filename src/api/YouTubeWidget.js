import axios from "axios";

export const getChannelTitleAndAvatar = async function(channelId, apiKey) {
    const response = await axios.get(`https://www.googleapis.com/youtube/v3/channels?part=snippet&id=${channelId}&key=${apiKey}`);

    return response.data;
};

export const getSubscribersCount = async function(channelId, apiKey) {
    const response = await axios.get(`https://www.googleapis.com/youtube/v3/channels?part=statistics&id=${channelId}&key=${apiKey}`);

    return response.data;
};

export const getVideos = async function(channelId, apiKey, resultsCount) {
    const response = await axios.get(`https://www.googleapis.com/youtube/v3/search?key=${apiKey}&channelId=${channelId}&part=snippet,id&order=date&maxResults=${resultsCount}`);

    return response.data;
};

export const loadMoreVideos = async function(channelId, apiKey, resultsCount, nextPageToken) {
    const response = await axios.get(`https://www.googleapis.com/youtube/v3/search?key=${apiKey}&channelId=${channelId}&part=snippet,id&order=date&pageToken=${nextPageToken}&maxResults=${resultsCount}`);

    return response.data;
};