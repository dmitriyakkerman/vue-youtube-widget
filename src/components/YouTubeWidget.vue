<template>
    <div class="youtube-widget">
        <div class="youtube-widget__header">
            <div class="youtube-widget__info">
                <img :src="channelAvatar" alt="" class="youtube-widget__logo">
                <div class="youtube-widget__data">
                    <h2 class="youtube-widget__title">{{ channelTitle }}</h2>
                    <div class="youtube-widget__subscribers-count">{{ Number(subscribersCount/1000).toFixed(subscribersCountToFixed) }} {{ subscribersCountText }}</div>
                </div>
            </div>
            <a :href="'http://www.youtube.com/channel/' + channelId + '?sub_confirmation=1'" class="youtube-widget__subscribe-btn">{{ subscribeBtnText }}</a>
        </div>
        <div class="youtube-widget__inner">
            <iframe v-for="video in videos" :key="video.id.videoId" :src="'https://www.youtube.com/embed/' + video.id.videoId" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
        </div>
        <button v-if="totalVideosCount > videos.length" class="youtube-widget__load-more-btn" @click="loadMore" :disabled="isSending">{{ loadMoreBtnText }}</button>
    </div>
</template>

<script>

    import { getChannelTitleAndAvatar, getSubscribersCount, getVideos, loadMoreVideos } from "../api/YouTubeWidget";

    export default {
        props: {
            apiKey: {
                type: String,
                required: true
            },
            channelId: {
                type: String,
                required: true
            },
            resultsPerRequest: {
                type: Number,
                required: false,
                default: 1
            },
            subscribersCountToFixed: {
                type: Number,
                required: false,
                default: 1
            },
            subscribersCountText: {
                type: String,
                required: false,
                default: 'subscribers'
            },
            subscribeBtnText: {
                type: String,
                required: false,
                default: 'Subscribe'
            },
            loadMoreBtnText: {
                type: String,
                default: 'Load more'
            }
        },
        data() {
            return {
                subscribersCount: null,
                channelAvatar: '',
                channelTitle: '',
                videos: [],
                totalVideosCount: null,
                nextPageToken: null,
                isSending: false
            }
        },
        methods: {
            async getChannelTitleAndAvatar() {
                this.isSending = true;

                const response = await getChannelTitleAndAvatar(this.channelId, this.apiKey);

                this.channelAvatar = response["items"][0].snippet.thumbnails.default.url;
                this.channelTitle = response["items"][0].snippet.localized.title;

                this.isSending = false;
            },
            async getSubscribersCount() {
                this.isSending = true;

                const response = await getSubscribersCount(this.channelId, this.apiKey);

                this.subscribersCount = response["items"][0].statistics.subscriberCount;
                this.totalVideosCount = response["items"][0].statistics.videoCount;

                this.isSending = false;
            },
            async getVideos() {
                this.isSending = true;

                const response = await getVideos(this.channelId, this.apiKey, this.resultsPerRequest);

                this.videos = response.items;
                this.nextPageToken = response.nextPageToken;

                this.isSending = false;
            },
            async loadMore() {
                this.isSending = true;

                const response = await loadMoreVideos(this.channelId, this.apiKey, this.resultsPerRequest, this.nextPageToken);

                this.videos = [...this.videos, ...response.items];
                this.nextPageToken = response.nextPageToken;

                this.isSending = false;
            },
        },
        mounted() {
            this.getChannelTitleAndAvatar();
            this.getSubscribersCount();
            this.getVideos();
        }
    }

</script>