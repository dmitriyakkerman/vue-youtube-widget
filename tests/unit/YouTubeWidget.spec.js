import { createLocalVue, mount } from '@vue/test-utils';
import YouTubeWidget from '../../src/components/YouTubeWidget.vue';

const mockData = {
    apiKey: 'yourapikey',
    channelId: 'UCLA_DiR1FfKNvjuUpBHmylQ',
    resultsPerRequest: 1,
    subscribersCountText: "подписчиков",
    subscribeBtnText: "Подписаться",
    loadMoreBtnText: "Загрузить еще",
    httpResponse: {
        data: {
            items: [
                {
                    id: {
                        videoId: '1'
                    },
                    snippet: {
                        thumbnails: {
                            default: {
                                url: 'avatar-url'
                            }
                        },
                        localized: {
                            title: 'NASA'
                        }
                    },
                    statistics: {
                        subscriberCount: 100,
                        videoCount: 100
                    }
                }
            ],
            nextPageToken: 'E7GFKSDGKLF22f'
        }
    }
};

jest.mock('axios', () => ({
    get: () => Promise.resolve(mockData.httpResponse)
}));

describe('YouTubeWidget.vue', () => {
    const vueInstance = createLocalVue();

    const wrapper = mount(YouTubeWidget, {
        vueInstance,
        propsData: {
            apiKey: mockData.apiKey,
            channelId: mockData.channelId,
            resultsPerRequest: mockData.resultsPerRequest,
            subscribersCountText: mockData.subscribersCountText,
            subscribeBtnText: mockData.subscribeBtnText,
            loadMoreBtnText: mockData.loadMoreBtnText
        }
    });

    it('initialized correctly', () => {
        expect(wrapper.isVueInstance()).toBe(true);
        expect(wrapper.is(YouTubeWidget)).toBe(true);
    });

    it('renders props when passed', () => {
        expect(wrapper.props().apiKey).toBe(mockData.apiKey);
        expect(wrapper.props().channelId).toBe(mockData.channelId);
        expect(wrapper.props().resultsPerRequest).toBe(mockData.resultsPerRequest);
        expect(wrapper.props().subscribersCountText).toBe(mockData.subscribersCountText);
        expect(wrapper.props().subscribeBtnText).toBe(mockData.subscribeBtnText);
        expect(wrapper.props().loadMoreBtnText).toBe(mockData.loadMoreBtnText);

        expect(wrapper.find('.youtube-widget__subscribe-btn').attributes('href')).toEqual('http://www.youtube.com/channel/' + mockData.channelId + '?sub_confirmation=1');
        expect(wrapper.find('.youtube-widget__subscribe-btn').text()).toEqual(mockData.subscribeBtnText);
        expect(wrapper.find('.youtube-widget__load-more-btn').text()).toEqual(mockData.loadMoreBtnText);
    });

    it("fetches channel avatar src", async () => {
        wrapper.vm.$nextTick(() => {
            expect(wrapper.vm.channelAvatar).toEqual(mockData.httpResponse.data["items"][0].snippet.thumbnails.default.url);
            expect(wrapper.find('.youtube-widget__logo').attributes('src')).toEqual(mockData.httpResponse.data["items"][0].snippet.thumbnails.default.url);
        })
    });

    it("fetches channel title", async () => {
        wrapper.vm.$nextTick(() => {
            expect(wrapper.vm.channelTitle).toEqual(mockData.httpResponse.data["items"][0].snippet.localized.title);
            expect(wrapper.find('.youtube-widget__title').text()).toEqual(mockData.httpResponse.data["items"][0].snippet.localized.title);
        })
    });

    it("fetches channel subscribers count", async () => {
        wrapper.vm.$nextTick(() => {
            expect(wrapper.vm.subscribersCount).toEqual(mockData.httpResponse.data["items"][0].statistics.subscriberCount);
            expect(wrapper.find('.youtube-widget__subscribers-count').text()).toEqual(Number(mockData.httpResponse.data["items"][0].statistics.subscriberCount) + ' ' + mockData.subscribersCountText);
        })
    });

    it("fetches 1 channel video when loaded", async () => {
        wrapper.vm.$nextTick(() => {
            expect(wrapper.vm.videos.length).toBe(1);
            expect(wrapper.vm.videos).toEqual(mockData.httpResponse.data.items);
            expect(wrapper.findAll('iframe').length).toBe(1);
        })
    });

    it('calls "loadMore" method after click on "Load More" button and fetches one more video', async () => {
        const loadMoreBtn = wrapper.find(".youtube-widget__load-more-btn");
        await loadMoreBtn.trigger("click");

        wrapper.vm.$nextTick(() => {
            expect(wrapper.vm.videos.length).toBe(2);
            expect(wrapper.findAll('iframe').length).toBe(2);
        })
    });
});
