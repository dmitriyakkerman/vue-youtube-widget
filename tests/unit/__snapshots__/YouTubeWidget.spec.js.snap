// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YouTubeWidget.vue should match snapshot 1`] = `
<div class="youtube-widget">
  <div class="youtube-widget__header">
    <div class="youtube-widget__info"><img src="avatar-url" alt="" class="youtube-widget__logo">
      <div class="youtube-widget__data">
        <h2 class="youtube-widget__title">NASA</h2>
        <div class="youtube-widget__subscribers-count">100 подписчиков</div>
      </div>
    </div> <a href="http://www.youtube.com/channel/UCLA_DiR1FfKNvjuUpBHmylQ?sub_confirmation=1" class="youtube-widget__subscribe-btn">Подписаться</a>
  </div>
  <div class="youtube-widget__inner"><iframe src="https://www.youtube.com/embed/1" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen="allowfullscreen"></iframe><iframe src="https://www.youtube.com/embed/1" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen="allowfullscreen"></iframe></div> <button class="youtube-widget__load-more-btn">Загрузить еще</button>
</div>
`;
